cage.c

```diff
@@ -49,26 +49,73 @@
 #include "xwayland.h"
 #endif
 
+static int
+sigchld_handler(int fd, uint32_t mask, void *data)
+{
+	struct wl_display *display = data;
+
+	/* Close Cage's read pipe. */
+	close(fd);
+
+	if (mask & WL_EVENT_HANGUP) {
+		wlr_log(WLR_DEBUG, "Child process closed normally");
+	} else if (mask & WL_EVENT_ERROR) {
+		wlr_log(WLR_DEBUG, "Connection closed by server");
+	}
+
+	wl_display_terminate(display);
+	return 0;
+}
+
 static bool
-spawn_primary_client(char *argv[], pid_t *pid_out)
+spawn_primary_client(struct wl_display *display, char *argv[], pid_t *pid_out, struct wl_event_source **sigchld_source)
 {
+	int fd[2];
+	if (pipe(fd) != 0) {
```

Comment:
By default, `pipe` creates FDs without the CLOEXEC flag set, which means they will be leaked to any other child process spawned. Would be nice to set the CLOEXEC flag to prevent the leak.